README.md
# [Geocoder GeoJSON](https://www.npmjs.com/package/geocoder-geojson)

[![Build Status](https://travis-ci.org/DenisCarriere/geocoder-geojson.svg?branch=master)](https://travis-ci.org/DenisCarriere/geocoder-geojson)
[![Circle CI](https://circleci.com/gh/DenisCarriere/geocoder-geojson.svg?style=svg)](https://circleci.com/gh/DenisCarriere/geocoder-geojson)
[![Coverage Status](https://coveralls.io/repos/github/DenisCarriere/geocoder-geojson/badge.svg?branch=master)](https://coveralls.io/github/DenisCarriere/geocoder-geojson?branch=master)
[![npm version](https://badge.fury.io/js/geocoder-geojson.svg)](https://badge.fury.io/js/geocoder-geojson)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/DenisCarriere/geocoder-geojson/master/LICENSE)

Geocoding results according to the [GeoJSON specification](http://geojson.org/geojson-spec.html).

## Install

```bash
$ npm install --save geocoder-geojson
```

Install globaly to access `geocode` via your command prompt.

```bash
$ npm install -g geocoder-geojson
$ geocode --version
$ geocode --help
```

## Quickstart

```javascript
import geocoder from 'geocoder-geojson'

geocoder.google('Ottawa, ON')
  .then(geojson => console.log(geojson))
```

## Supports

-   [Webpack](https://webpack.js.org/)
-   [Browserify](http://browserify.org/)
-   [ES6 (ES2015)](https://babeljs.io/learn-es2015/)
-   [NodeJS 6 & 7](https://nodejs.org/en/)

## Features

| Name                            | Coverage | Restrictions                        |
| ------------------------------- | :------- | :---------------------------------- |
| [google](#google)               | Global   | Free & API Key - RateLimit 2500/day |
| [googleReverse](#googlereverse) | Global   | Free & API Key - RateLimit 2500/day |
| [mapbox](#mapbox)               | Global   | API Key                             |
| [mapboxReverse](#mapboxreverse) | Global   | API Key                             |
| [bing](#bing)                   | Global   | API Key                             |
| [wikdata](#wikidata)            | Global   | Free                                |

## CLI

```bash
$ geocode --provider bing "Ottawa ON"
$ geocode -p wikidata --nearest [-75.7,45.4] Ottawa
$ geocode -p google --limit 3 "Ottawa ON"
```

Using `jq` to filter JSON data

```bash
$ geocode -p wikidata --nearest [-75.7,45.4] Ottawa | jq .features[0].id
"Q1930"
```

## Roadmap

-   Implement all geocoder providers from [`Python Geocoder`](https://github.com/DenisCarriere/geocoder)

## API

### mapbox

Mapbox Provider

<https://www.mapbox.com/api-documentation/#geocoding>

**Parameters**

-   `address` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** Location for your search
-   `options` **\[MapboxOptions]** Mapbox Options
    -   `options.access_token` **\[[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)]** Access token or environment variable `MAPBOX_ACCESS_TOKEN`
    -   `options.mode` **\[[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)]** Mode mapbox.places or mapbox.places-permanent (optional, default `'mapbox.places'`)
    -   `options.country` **\[[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)]** ISO 3166 alpha 2 country codes, separated by commas
    -   `options.proximity` **\[LngLat]** Location around which to bias results, given as longitude,latitude
    -   `options.types` **\[[Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)>]** Filter results by one or more type.
    -   `options.autocomplete` **\[[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)]** Whether or not to return autocomplete results. (optional, default `true`)
    -   `options.bbox` **\[BBox]** Bounding box within which to limit results, given as minX,minY,maxX,maxY
    -   `options.limit` **\[[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)]** Limit the number of results returned. (optional, default `5`)

**Examples**

```javascript
const geojson = await geocoder.mapbox('Ottawa, ON')
```

Returns **[Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;Points>** GeoJSON Point FeatureCollection

### mapboxReverse

Mapbox Provider (Reverse)

<https://www.mapbox.com/api-documentation/#geocoding>

**Parameters**

-   `lnglat` **LngLat** Longitude & Latitude [x, y]
-   `options` **\[MapboxOptions]** Mapbox Options
    -   `options.access_token` **\[[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)]** Access token or environment variable `MAPBOX_ACCESS_TOKEN`
    -   `options.mode` **\[[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)]** Mode mapbox.places or mapbox.places-permanent (optional, default `'mapbox.places'`)
    -   `options.country` **\[[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)]** ISO 3166 alpha 2 country codes, separated by commas
    -   `options.proximity` **\[LngLat]** Location around which to bias results, given as longitude,latitude
    -   `options.types` **\[[Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)>]** Filter results by one or more type.
    -   `options.autocomplete` **\[[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)]** Whether or not to return autocomplete results. (optional, default `true`)
    -   `options.bbox` **\[BBox]** Bounding box within which to limit results, given as minX,minY,maxX,maxY
    -   `options.limit` **\[[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)]** Limit the number of results returned. (optional, default `1`)

**Examples**

```javascript
const geojson = await geocoder.mapbox('Ottawa, ON')
```

Returns **[Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;Points>** GeoJSON Point FeatureCollection

### google

Google Provider

<https://developers.google.com/maps/documentation/geocoding>

**Parameters**

-   `address` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** Location for your search
-   `options` **\[GoogleOptions]** Google Options
    -   `options.language` **\[[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)]** The language in which to return results (optional, default `en`)
    -   `options.short` **\[[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)]** Address components have long or short results (optional, default `false`)

**Examples**

```javascript
const geojson = await geocoder.google('Ottawa, ON')
```

Returns **[Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;Points>** GeoJSON Point FeatureCollection

### googleReverse

Google Provider (Reverse)

<https://developers.google.com/maps/documentation/geocoding>

**Parameters**

-   `lnglat` **LngLat** Longitude & Latitude [x, y]
-   `options` **\[GoogleOptions]** Google Options
    -   `options.language` **\[[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)]** The language in which to return results (optional, default `en`)
    -   `options.short` **\[[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)]** Address components have long or short results (optional, default `false`)

**Examples**

```javascript
const geojson = await geocoder.googleReverse([-75.1, 45.1])
```

Returns **[Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;Points>** GeoJSON Point FeatureCollection

### bing

Bing Provider

<https://msdn.microsoft.com/en-us/library/ff701714.aspx>

**Parameters**

-   `address` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** Location for your search
-   `options` **\[BingOptions]** Bing Options
    -   `options.key` **\[[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)]** API key or environment variable `BING_API_KEY`
    -   `options.maxResults` **\[[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)]** Specifies the maximum number of locations to return in the response.

**Examples**

```javascript
const geojson = await geocoder.bing('Ottawa, ON')
```

Returns **[Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;Points>** GeoJSON Point FeatureCollection

### wikidata

Wikidata Provider

<https://query.wikidata.org/>

**Parameters**

-   `address` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** Location for your search
-   `options` **\[Options]** Wikidata Options
    -   `options.nearest` **\[LngLat]** Nearest location from a given LngLat
    -   `options.radius` **\[[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)]** Maximum radius from nearest LngLat
    -   `options.languages` **\[[Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)>]** Exact match on a list of languages
    -   `options.subclasses` **\[[Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)>]** Filter results by Wikidata subclasses

**Examples**

```javascript
const geojson = await geocoder.wikidata('Ottawa')
```

Returns **[Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;Points>** GeoJSON Point FeatureCollection

### request

Generic GET function to normalize all of the requests

**Parameters**

-   `url` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** URL
-   `geojsonParser` **[function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function)** Customized function to generate a GeoJSON Point FeatureCollection
-   `params` **\[[Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)](default {})** Query String
-   `options` **\[[Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)](default utils.Options)** Options used for HTTP request & GeoJSON Parser function

Returns **[Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;Points>** Results in GeoJSON FeatureCollection Points


CHANGELOG.md
# Changelog

## 2.0.0 -2017-03-4

- Refactor entire code to ES6 (removed Typescript, only keeping .d.ts)

## 1.10.0 - 2017-01-17

-   Updated docs
-   Removed `colors` dependency

## 1.9.0 - 2016-12-3

-   Replace all test cases with `Jest` - Major refactoring
-   Added more Mapbox Options
-   Added ISO_3166 & ISO_639 codes to `utils`

## 1.8.0 - 2016-11-16

-   Implemented Wikidata SPARQL

## 1.7.0 - 2016-11-11

-   Add Command Line Interface (CLI)
-   Add Wikidata provider
-   Add `--nearest` to Wikidata
-   Add wikidata `places` precision
-   Add `places` to filter out `places=*` when using `nearest`
-   Seperate tests per providers `test/test.<provider>.ts`

## 1.6.0 - 2016-11-3

-   Replace `isomorphic-fetch` & `qs` with `axios` (less headaches)
-   Deprecate `replaceStreetSuffix` function

## 1.5.2 - 2016-10-26

-   Add `@types` to dependencies
-   Replaced `request` dependency for `isomorphic-fetch`.

## 1.4.0 - 2016-10-22

-   Replaced `@turf/turf` for namespaced modules `@turf/helpers`, `@turf/distance` & `@turf/bbox-polygon`
-   Removed `lodash` from dependencies
-   Change docs to `await` instead of `.then()`
-   Add default export
-   Update JSDocs options

## 1.2.1 - 2016-10-20

-   Remove OSM attribute creation from Google & Bing
-   Validatation of LngLat - Common mistake is rerversing the coordinates to LatLng
-   Added Mapbox & Mapbox Reverse providers
-   Added generic GET function to normalize all of the requests

```javascript
// Environment Variable MAPBOX_API_KEY
geocoder.mapbox('Ottawa, ON')
geocoder.mapboxReverse([-75.1, 45.1])
```

## 1.1.0 - 2016-10-17

-   Added Google Reverse method
-   Changed default Google option `short=false`

```javascript
geocoder.googleReverse([-75.1, 45.1])
```

## 1.0.0 - 2016-10-14

First Stable release was created

Features:

-   `google` provider
-   `confidenceScore`
-   `replaceStreetSuffix`


index.ts
import axios from 'axios'
import * as utils from './utils'
import { LngLat, error, GeoJSONParser } from './utils'
import * as iso3166 from './utils/ISO_3166-1_alpha-2'
import {
  Bing,
  Google,
  Mapbox,
  Wikidata } from './providers'

export type Providers = 'bing' | 'google' | 'mapbox' | 'mapboxReverse' | 'wikidata' | 'googleReverse' | 'mapboxReverse'
export type Points = GeoJSON.FeatureCollection<GeoJSON.Point>

/**
 * Mapbox Provider
 *
 * https://www.mapbox.com/api-documentation/#geocoding
 *
 * @param {string} address Location for your search
 * @param {MapboxOptions} [options] Mapbox Options
 * @param {string} [options.access_token] Access token or environment variable `MAPBOX_ACCESS_TOKEN`
 * @param {string} [options.mode='mapbox.places'] Mode mapbox.places or mapbox.places-permanent
 * @param {string} [options.country] ISO 3166 alpha 2 country codes, separated by commas
 * @param {LngLat} [options.proximity] Location around which to bias results, given as longitude,latitude
 * @param {Array<string>} [options.types] Filter results by one or more type.
 * @param {boolean} [options.autocomplete=true] Whether or not to return autocomplete results.
 * @param {BBox} [options.bbox] Bounding box within which to limit results, given as minX,minY,maxX,maxY
 * @param {number} [options.limit=5] Limit the number of results returned.
 * @returns {Promise<Points>} GeoJSON Point FeatureCollection
 * @example
 * const geojson = await geocoder.mapbox('Ottawa, ON')
 */
export function mapbox(address: string, options = Mapbox.Options): Promise<Points> {
  // Define options
  const mode = options.mode || Mapbox.Options.mode
  const access_token = options.access_token || options.key || process.env.MAPBOX_ACCESS_TOKEN
  const country = options.country
  const limit = options.limit || Mapbox.Options.limit

  // Validation
  if (!access_token) { error('[options.access_token] is required') }
  if (['mapbox.places', 'mapbox.places-permanent'].indexOf(mode) === -1) { error('--mode is invalid') }
  if (country && iso3166.codes[country] === undefined) { error('--country is invalid') }

  // URL Parameters
  const params = {
    access_token,
    country,
    limit,
  }

  // Request
  const url = `https://api.mapbox.com/geocoding/v5/${mode}/${address}.json`
  return request(url, Mapbox.toGeoJSON, params, options)
}

/**
 * Mapbox Provider (Reverse)
 *
 * https://www.mapbox.com/api-documentation/#geocoding
 *
 * @param {LngLat} lnglat Longitude & Latitude [x, y]
 * @param {MapboxOptions} [options] Mapbox Options
 * @param {string} [options.access_token] Access token or environment variable `MAPBOX_ACCESS_TOKEN`
 * @param {string} [options.mode='mapbox.places'] Mode mapbox.places or mapbox.places-permanent
 * @param {string} [options.country] ISO 3166 alpha 2 country codes, separated by commas
 * @param {LngLat} [options.proximity] Location around which to bias results, given as longitude,latitude
 * @param {Array<string>} [options.types] Filter results by one or more type.
 * @param {boolean} [options.autocomplete=true] Whether or not to return autocomplete results.
 * @param {BBox} [options.bbox] Bounding box within which to limit results, given as minX,minY,maxX,maxY
 * @param {number} [options.limit=1] Limit the number of results returned.
 * @returns {Promise<Points>} GeoJSON Point FeatureCollection
 * @example
 * const geojson = await geocoder.mapbox('Ottawa, ON')
 */
export function mapboxReverse(lnglat: string | LngLat, options = Mapbox.Options): Promise<Points> {
  // Define options
  const mode = options.mode || Mapbox.Options.mode
  const access_token = options.access_token || options.key || process.env.MAPBOX_ACCESS_TOKEN
  const country = options.country
  const limit = options.limit || Mapbox.Options.limit

  // Validation
  if (['mapbox.places', 'mapbox.places-permanent'].indexOf(mode) > 0) { error('--mode is invalid') }
  lnglat = utils.validateLngLat(lnglat)
  if (!access_token) { error('--access_token is required') }

  // URL Parameters
  const params = {
    access_token,
    country,
    limit,
  }

  // Request
  const url = `https://api.mapbox.com/geocoding/v5/${mode}/${lnglat.join(',')}.json`
  return request(url, Mapbox.toGeoJSON, params, options)
}

/**
 * Google Provider
 *
 * https://developers.google.com/maps/documentation/geocoding
 *
 * @param {string} address Location for your search
 * @param {GoogleOptions} [options] Google Options
 * @param {string} [options.language=en] The language in which to return results
 * @param {boolean} [options.short=false] Address components have long or short results
 * @returns {Promise<Points>} GeoJSON Point FeatureCollection
 * @example
 * const geojson = await geocoder.google('Ottawa, ON')
 */
export function google(address: string, options = Google.Options): Promise<Points> {
  // Define Options
  const language = options.language || Google.Options.language
  const sensor = options.sensor || Google.Options.sensor
  options.short = options.short || Google.Options.short

  // URL Parameters
  const params = {
    address,
    language,
    sensor,
  }

  // Request
  const url = 'https://maps.googleapis.com/maps/api/geocode/json'
  return request(url, Google.toGeoJSON, params, options)
}

/**
 * Google Provider (Reverse)
 *
 * https://developers.google.com/maps/documentation/geocoding
 *
 * @param {LngLat} lnglat Longitude & Latitude [x, y]
 * @param {GoogleOptions} [options] Google Options
 * @param {string} [options.language=en] The language in which to return results
 * @param {boolean} [options.short=false] Address components have long or short results
 * @returns {Promise<Points>} GeoJSON Point FeatureCollection
 * @example
 * const geojson = await geocoder.googleReverse([-75.1, 45.1])
 */
export function googleReverse(lnglat: string | LngLat, options = Google.Options): Promise<Points> {
  // Define Options
  options.language = options.language || Google.Options.language
  options.sensor = options.sensor || Google.Options.sensor
  options.short = options.short || Google.Options.short
  const [lng, lat] = utils.validateLngLat(lnglat)

  // URL Parameters
  const params = {
    address: [lat, lng].join(','),
    sensor: options.sensor,
  }

  // Request
  const url = 'https://maps.googleapis.com/maps/api/geocode/json'
  return request(url, Google.toGeoJSON, params, options)
}

/**
 * Bing Provider
 *
 * https://msdn.microsoft.com/en-us/library/ff701714.aspx
 *
 * @param {string} address Location for your search
 * @param {BingOptions} [options] Bing Options
 * @param {string} [options.key] API key or environment variable `BING_API_KEY`
 * @param {string} [options.maxResults] Specifies the maximum number of locations to return in the response.
 * @returns {Promise<Points>} GeoJSON Point FeatureCollection
 * @example
 * const geojson = await geocoder.bing('Ottawa, ON')
 */
export function bing(address: string, options = Bing.Options): Promise<Points> {
  // Define Options
  const key = options.key || process.env.BING_API_KEY
  const maxResults = options.maxResults || options.limit || Bing.Options.maxResults

  // Validation
  if (!key) { error('--key is required') }

  // URL Parameters
  const params = {
    inclnb: 1,
    key,
    o: 'json',
    q: address,
    maxResults,
  }

  // Request
  const url = 'http://dev.virtualearth.net/REST/v1/Locations'
  return request(url, Bing.toGeoJSON, params, options)
}

/**
 * Wikidata Provider
 *
 * https://query.wikidata.org/
 *
 * @param {string} address Location for your search
 * @param {Options} [options] Wikidata Options
 * @param {LngLat} [options.nearest] Nearest location from a given LngLat
 * @param {number} [options.radius] Maximum radius from nearest LngLat
 * @param {Array<string>} [options.languages] Exact match on a list of languages
 * @param {Array<string>} [options.subclasses] Filter results by Wikidata subclasses
 * @returns {Promise<Points>} GeoJSON Point FeatureCollection
 * @example
 * const geojson = await geocoder.wikidata('Ottawa')
 */
export function wikidata(address: string, options = Wikidata.Options): Promise<Points> {
  // Define Options
  options.subclasses = options.subclasses || Wikidata.Options.subclasses
  options.languages = options.languages || Wikidata.Options.languages
  options.radius = options.radius || Wikidata.Options.radius
  options.sparql = options.sparql || Wikidata.Options.sparql

  // Validation
  if (!options.nearest) { error('--nearest is required') }

  // SPARQL Query
  const query = Wikidata.createQuery(address, options)

  // URL Parameters
  const params = {
    query,
  }

  // Request
  const url = 'https://query.wikidata.org/bigdata/namespace/wdq/sparql'
  return request(url, Wikidata.toGeoJSON, params, options)
}

/**
 * Generic GET function to normalize all of the requests
 *
 * @param {string} url URL
 * @param {function} geojsonParser Customized function to generate a GeoJSON Point FeatureCollection
 * @param {Object} params Query String
 * @param {Object} options Options used for HTTP request & GeoJSON Parser function
 * @returns {Promise<Points>} Results in GeoJSON FeatureCollection Points
 */
export async function request(url: string, geojsonParser: GeoJSONParser, params = {}, options = utils.Options): Promise<Points> {
  // Create custom Axios instance
  const instance = axios.create({})

  // Remove any existing default Authorization headers
  if (instance.defaults.headers.common && instance.defaults.headers.common.Authorization) { delete instance.defaults.headers.common.Authorization }
  if (instance.defaults.headers.Authorization) { delete instance.defaults.headers.Authorization }

  // Handle request
  const response = await instance.get(url, {params})
  if (options.raw !== undefined) { return response.data }
  const geojson = geojsonParser(response.data, options)
  return geojson
}

export function get(provider: Providers, query: string, options?: any): Promise<Points> {
  if (provider === 'bing') { return bing(query, options) }
  if (provider === 'google') { return google(query, options) }
  if (provider === 'mapbox') { return mapbox(query, options) }
  if (provider === 'wikidata') { return wikidata(query, options) }
  if (provider === 'googleReverse') { return googleReverse(query, options) }
  if (provider === 'mapboxReverse') { return mapboxReverse(query, options) }
}
